dlfcn.c

```diff
@@ -208,6 +208,157 @@ static BOOL MyEnumProcessModules( HANDLE hProcess, HMODULE *lphModule, DWORD cb,
     return EnumProcessModulesPtr( hProcess, lphModule, cb, lpcbNeeded );
 }
 
+// Create a converted multibyte character string from a wide character string.
+static char *utf8_encode( const wchar_t *wstr )
+{
+    size_t input_str_len = wcslen( wstr );
+    if ( !wstr || input_str_len == 0 )
+    {
+        return NULL;
+    }
+    int size_needed = WideCharToMultiByte( CP_UTF8, 0, wstr, input_str_len, NULL, 0, NULL, NULL );
+    char *to_return = (char *) malloc( size_needed * sizeof( char ) );
+    if ( !to_return )
+        return NULL;
+    WideCharToMultiByte( CP_UTF8, 0, wstr, input_str_len, to_return, size_needed, NULL, NULL );
+    return to_return;
+}
+
+// Create a converted wide character string from a multibyte character string.
+static wchar_t *utf8_decode( const char *str )
+{
+    size_t input_str_len = strlen( str );
+    if ( !str || input_str_len == 0 )
+    {
+        return NULL;
+    }
+    int size_needed = MultiByteToWideChar( CP_UTF8, 0, str, input_str_len, NULL, 0 );
+    wchar_t *to_return = (wchar_t *) malloc( size_needed * sizeof( wchar_t ) );
+    if ( !to_return )
+        return NULL;
+    MultiByteToWideChar( CP_UTF8, 0, str, input_str_len, to_return, size_needed );
+    return to_return;
+}
+
+void *wdlopen( const wchar_t *file, int mode )
+{
+    HMODULE hModule;
+    UINT uMode;
+
+    error_occurred = FALSE;
+
+    /* Do not let Windows display the critical-error-handler message box */
+    uMode = SetErrorMode( SEM_FAILCRITICALERRORS );
+
+    if( file == 0 )
+    {
+        /* POSIX says that if the value of file is 0, a handle on a global
+         * symbol object must be provided. That object must be able to access
+         * all symbols from the original program file, and any objects loaded
+         * with the RTLD_GLOBAL flag.
+         * The return value from GetModuleHandle( ) allows us to retrieve
+         * symbols only from the original program file. EnumProcessModules() is
+         * used to access symbols from other libraries. For objects loaded
+         * with the RTLD_LOCAL flag, we create our own list later on. They are
+         * excluded from EnumProcessModules() iteration.
+         */
+        hModule = GetModuleHandle( NULL );
+
+        if( !hModule )
+            save_err_str( "(null)" );
+    }
+    else
+    {
+        HANDLE hCurrentProc;
+        DWORD dwProcModsBefore, dwProcModsAfter;
+        wchar_t lpFileName[MAX_PATH];
+        size_t i, len;
+
+        len = wcslen( file );
+
+        if( len >= sizeof( lpFileName ) )
+        {
+            SetLastError( ERROR_FILENAME_EXCED_RANGE );
+            char* fileMultibyte = utf8_encode( file );
+            if ( !fileMultibyte )
+                return NULL;
+            save_err_str( fileMultibyte );

```

Comment:
This is wrong. Windows -A functions do not use UTF-8 but rather encoding according to current code page. Therefore also `save_err_str` function expects string according to current code page and not UTF-8. Incorrect encoding would result in garbaged output.

Reply:
I wanted to do a minimal change to the code, and since `save_err_str` is only to add error messages, it would be beyond the scope of the change, which is to add `wchar_t` input functionality.

Reply:
@pali I am confused here, design-wise. How would I have a wide version of `save_err_str`, while in the same time, when calling `dlerror` return strings from both? or do you think it's better to have another buffer, say, `wdlerror`, and I wouldn't need to convert the string in `wdlopen` when saving the error string?

Reply:
You cannot convert arbitrary UNICODE string (wchar_t is UTF-16 on WIN32) to multibyte string according to current code page. That is the fact.

Reply:
Yes, but don't you think it's a good idea to create a version of this function that takes a wide character, much like most functions in the Windows API, where there are functions that accept/return char(\*) another ones that accept/return wchar_t(\*)?

Reply:
I do not know what if it is a good idea or not... maybe it needs wdlerror? I have not thinging about it yet.

But in any case, current solution with UTF-8 is broken.